### PR TITLE
Add the possibility to write Golang function as Golang-native HTTP handlers

### PR DESCRIPTION
**Pull request by markusthoemmes**
This enables functions to take the form `func(http.ResponseWriter, *http.Request)` or be of the `http.Handler` type as is the Golang-native way of writing HTTP handlers in the standard library. It does that by translating a request as produced by OpenWhisk into a `*http.Request` and recording the respective response in-memory. It too will then be translated into a format OpenWhisk understand and is then returned.

This model really only makes sense with `web: raw` enabled (see analysis below on why).

### http-handler-as-a-function

```golang
package main

import (
	"fmt"
	"net/http"
)

var Main = http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
	fmt.Fprint(rw, "Hello HTTP")
})
```

### gorilla-mux-as-a-function with off-the-shelf middleware

```golang
package main

import (
	"fmt"
	"net/http"
	"os"

	"github.com/gorilla/handlers"
	"github.com/gorilla/mux"
)

var Main http.Handler

func init() {
	m := mux.NewRouter()
	m.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprint(w, "hello base!")
	})
	m.HandleFunc("/foo", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprint(w, "hello /foo")
	})
	m.HandleFunc("/bar", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprint(w, "hello /bar")
	})
	m.HandleFunc("/hello/{name}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		fmt.Fprintf(w, "hello %s", vars["name"])
	})

	Main = handlers.LoggingHandler(os.Stdout, m)
}

```

### CODE REVIEW

**Comment by markusthoemmes** [comment]
Note to self: This isn't done! It's somewhat ready for raw-http functions, but normal web actions don't quite work as well, because there's no way to get to the query parameters currently. I have to do a little more testing around that and see if it's safely feasible to infer where to put these. Initial idea/questions:
- What happens if I post a JSON body? 
- Could we throw the query into the body as a JSON?

**Comment by markusthoemmes** [comment]
# Test behavior

## Raw web action invoke with query and JSON

```
$ curl "https://$OW_API_HOST/api/v1/web/nimbella/default/tester/hello/world?test=foo" -H 'Content-Type: application/json' -d '{"name":"Jane"}'
{
  "__ow_body": "{\"name\":\"Jane\"}",
  "__ow_headers": { REDACTED },
  "__ow_method": "post",
  "__ow_path": "/hello/world",
  "__ow_query": "test=foo",
  "obj": {
    "testname": "testvalue"
  }
}
```

This matches the model nicely. Everything is where it "should" be and nothing has to be moved around to match the model. Almost...

Notice `obj`, which is a predefined parameter. How would we pass that into the function? Maybe through the context? (Requests have contexts).

## Normal web action invoke with query and JSON

```
$ curl "https://$OW_API_HOST/api/v1/web/nimbella/default/tester/hello/world?test=foo" -H 'Content-Type: application/json' -d '{"name":"Jane"}'
{
  "__ow_headers": { REDACTED },
  "__ow_method": "post",
  "__ow_path": "/hello/world",
  "name": "Jane",
  "obj": {
    "testname": "testvalue"
  },
  "test": "foo"
}
```

No __ow_body, the values are thrown together: We should extract them and pass them to the function. But how? As query or as body? We can't decide which comes from where.

## Normal web action invoke with query and non JSON

```
$ curl "https://$OW_API_HOST/api/v1/web/nimbella/default/tester/hello/world?test=foo" -H 'Content-Type: text/plain' -d 'foobar'
{
  "__ow_body": "foobar",
  "__ow_headers": { REDACTED },
  "__ow_method": "post",
  "__ow_path": "/hello/world",
  "obj": {
    "testname": "testvalue"
  },
  "test": "foo"
}
```

Both `__ow_body` and the extra values are passed along. Now we have only one choice: Pass the body as the body and pass the remaining non __ow values as query. Or, as mentioned above, provide extra values through the request's context.
IMO that destroys almost all the niceness of this model though and thus we'd be better of writing web actions in a "standard" way.

**Comment by markusthoemmes** [comment]
Bringing this back, I think we can make it work for web=raw invokes at least.

**Comment by markusthoemmes** [comment]
Note: This lacks auto-base64decode. We are able to detect this now.

File: golang1.17/lib/launcher.go [hunk 1]
```diff
@@ -278,3 +294,99 @@ func valueToError(val reflect.Value) error {
...
+		return nil, fmt.Errorf("failed to parse request: %w", err)
+	}
+	// Transform it to *http.Request
+	url, err := url.Parse("http://")
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %w", err)
+	}
+	url.Path = r.Path
+	url.RawQuery = r.Query
+	req, err := http.NewRequestWithContext(ctx, r.Method, url.String(), strings.NewReader(r.Body))
```
Comment (markusthoemmes) [inline golang1.17/lib/launcher.go]: TODO: base64decode automatically.

**Review by ElanHasson — APPROVED** [review]

**Pull request closed without merging**